File: Cpp/STL/290.py
# 290. Word Pattern

pattern = "abba"
s = "dog dog dog dog"
s = s.split(' ')
d = {}

if (len(set(pattern)) != len(set(s))):
    print(False)    # return
if (len(pattern) != len(s)):
    print(False)    # return
for i in range(len(pattern)):
    if pattern[i] not in d:
        d[pattern[i]] = s[i]
    elif d[pattern[i]] == s[i]:
        continue
    else:
        print(False)    #return
print(True)     #return

# My answer - using two dict

# pattern = "abba"
# s = "dog dog dog dog"
# arr = s.split(' ')
# ans = {}
# flag=0
# if len(pattern) != len(arr):
#     flag=1
# else:
#     for i in range(len(pattern)):
#         if pattern[i] in ans.keys():
#             if ans[pattern[i]] != arr[i]:
#                 flag=1
#                 break
#         else:
#             ans[pattern[i]] = arr[i]
# dup = {}
# for key, val in ans.items():
#     if val not in dup:
#         dup[val]=key
#     else:
#         flag=1
#         break
# if (flag==1):
#     print(False)
# else:
#     print(True)
    

File: Cpp/STL/290.cpp
#include <bits/stdc++.h>
using namespace std;

bool wordPattern(string pattern, string s)
{
    vector<string> arr;
    string t = "";
    for (int i = 0; i < s.length(); i++)
    {
        if (s[i] == ' ')
        {
            arr.push_back(t);
            t.clear();
        }
        else
        {
            t += s[i];
        }
    }
    arr.push_back(t);
    set<char> p(pattern.begin(), pattern.end());
    set<string> strarr(arr.begin(), arr.end());

    if (p.size() != strarr.size())
        return false;
    if (pattern.size() != arr.size())
        return false;

    map<char, string> m;
    for (int i = 0; i < pattern.size(); i++)
    {
        if (m.find(pattern[i]) == m.end())
            m[pattern[i]] = arr[i];
        else if (m[pattern[i]] == arr[i])
            continue;
        else
            return false;
    }

    return true;
}

int main()
{
    string pattern = "abba", s = "dog cat cat dog";
    cout << wordPattern(pattern, s);
}